### PR TITLE
EIP-1559: Fee market change for ETH 1.0 chain

### PR DESCRIPTION
**Pull request by econoar**
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-X.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your Github username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

### CODE REVIEW

**Comment by bmann** [comment]
This PR closes https://github.com/ethereum/EIPs/issues/1559

File: EIPS/eip-1559.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+An important aspect of this upgraded fee system is that miners only get to keep the tips. The BASEFEE is always burned (i.e. it is destroyed by the protocol). Burning this is important because it prevents miners from manipulating the fee in order to extract more fees from users. It also ensures that only ETH can ever be used to pay for transactions on Ethereum, cementing the economic value of ETH within the Ethereum platform.
+
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+**Parameters**
+* `FORK_BLKNUM`: TBD
+* `MINFEE_MAX_CHANGE_DENOMINATOR`: 8
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
* `BASEFEE_MAX_CHANGE_DENOMINATOR`: 8
```

Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
* `BASEFEE_MAX_CHANGE_DENOMINATOR`: 8
```

File: EIPS/eip-1559.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+* `SLACK_COEFFICIENT`: 3
+* `TARGET_GASUSED`: 8,000,000
+
+
+**Proposal**
+For all blocks where `block.number >= FORK_BLKNUM`:
+
+* Impose a hard in-protocol gas limit of `SLACK_COEFFICIENT * TARGET_GASUSED`, used instead of the gas limit calculated using the previously existing formulas
+* Replace the `GASLIMIT` field in the block header with a MINFEE field (the same field can be used)
+* Let `PARENT_MINFEE` be the parent block's `MINFEE` (or 1 billion wei if `block.number == FORK_BLKNUM`). A valid `MINFEE` is one such that `abs(MINFEE - PARENT_MINFEE) <= max(1, PARENT_MINFEE // MINFEE_MAX_CHANGE_DENOMINATOR)`
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
* Let `PARENT_BASEFEE` be the parent block's `BASEFEE` (or 1 billion wei if `block.number == FORK_BLKNUM`). A valid `BASEFEE` is one such that `abs(BASEFEE - PARENT_BASEFEE) <= max(1, PARENT_BASEFEE // BASEFEE_MAX_CHANGE_DENOMINATOR)`
```

File: EIPS/eip-1559.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+## Simple Summary
+<!--"If you can't explain it simply, you don't understand it well enough." Provide a simplified and layman-accessible explanation of the EIP.-->
+The current "first price auction" fee model in Ethereum is inefficient and needlessly costly to users. This EIP proposes a way to replace this with a mechanism that adjusts a base network fee based on network demand, creating better fee price efficiency and reducing the complexity of client software needed to avoid paying unnecessarily high fees.
+
+## Abstract
+<!--A short (~200 word) description of the technical issue being addressed.-->
+There is a MINFEE value in protocol, which can move up or down by a maximum of 1/8 in each block; initially, miners adjust this value to target an average gas usage of 8 million, increasing MINFEE if usage is higher and decreasing it if usage is lower. Transaction senders specify their fees by providing two values:
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
There is a BASEFEE value in protocol, which can move up or down by a maximum of 1/8 in each block; initially, miners adjust this value to target an average gas usage of 8 million, increasing BASEFEE if usage is higher and decreasing it if usage is lower. Transaction senders specify their fees by providing two values:
```

File: EIPS/eip-1559.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+* A "premium" gasprice which gets added onto the MINFEE gasprice, which can either be set to a fairly low value (eg. 1 gwei) to compensate miners for uncle rate risk or to a high value to compete during sudden bursts of activity. The MINFEE gets burned, the premium is given to the miner.
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
* A "premium" gasprice which gets added onto the BASEFEE gasprice, which can either be set to a fairly low value (eg. 1 gwei) to compensate miners for uncle rate risk or to a high value to compete during sudden bursts of activity. The BASEFEE gets burned, the premium is given to the miner.
```

File: EIPS/eip-1559.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+* `MINFEE_MAX_CHANGE_DENOMINATOR`: 8
+* `SLACK_COEFFICIENT`: 3
+* `TARGET_GASUSED`: 8,000,000
+
+
+**Proposal**
+For all blocks where `block.number >= FORK_BLKNUM`:
+
+* Impose a hard in-protocol gas limit of `SLACK_COEFFICIENT * TARGET_GASUSED`, used instead of the gas limit calculated using the previously existing formulas
+* Replace the `GASLIMIT` field in the block header with a MINFEE field (the same field can be used)
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
* Replace the `GASLIMIT` field in the block header with a BASEFEE field (the same field can be used)
```

File: EIPS/eip-1559.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+* Let `PARENT_MINFEE` be the parent block's `MINFEE` (or 1 billion wei if `block.number == FORK_BLKNUM`). A valid `MINFEE` is one such that `abs(MINFEE - PARENT_MINFEE) <= max(1, PARENT_MINFEE // MINFEE_MAX_CHANGE_DENOMINATOR)`
+* Redefine the way the `tx.gasprice` field is used: define `tx.fee_premium = tx.gasprice // 2**128` and `tx.fee_cap = tx.gasprice % 2**128`
+* During transaction execution, we calculate the cost to the `tx.origin` and the gain to the `block.coinbase` as follows:
+  * Let `gasprice = min(BASEFEE + tx.fee_premium, tx.fee_cap)`. The `tx.origin` initially pays `gasprice * tx.gas`, and gets refunded `gasprice * (tx.gas - gasused)`.
+  * The `block.coinbase` gains `(gasprice - BASEFEE) * gasused`.
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
  * The `block.coinbase` gains `(gasprice - BASEFEE) * gasused`. If `gasprice < BASEFEE` (due to the `fee_cap`), this means that the `block.coinbase` _loses_ funds from this operation; in this case, check that the post-balance is non-negative and throw an exception if it is negative.
```

File: EIPS/eip-1559.md [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+As a default strategy, miners set `BASEFEE` as follows. Let `delta = block.gas_used - TARGET_GASUSED` (possibly negative). Set `BASEFEE = PARENT_BASEFEE + PARENT_BASEFEE * delta // TARGET_GASUSED // BASEFEE_MAX_CHANGE_DENOMINATOR`, clamping this result inside of the allowable bounds if needed (with the parameter setting above clamping will not be required).
+
+## Backwards Compatibility
+<!--All EIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The EIP must explain how the author proposes to deal with these incompatibilities. EIP submissions without a sufficient backwards compatibility treatise may be rejected outright.-->
```
Comment (vbuterin) [inline EIPS/eip-1559.md]:
```suggestion
Transactions published before this EIP or by wallets that do not support this EIP will be interpreted by the above formulas as having a `fee_premium` of zero and a `fee_cap` of the fee that they submit. Provided that at least some miners are temporarily willing to be altruistic and accept zero-fee-premium transactions for a short period of time after the fork, this should not greatly affect usability. There is an invariant that a `gasprice` constructed "the old way" still constitutes an upper bound on the amount that a user will pay.
```

**Pull request merged**